ggml-cuda.cu

```diff
@@ -131,24 +131,83 @@ static __global__ void dequantize_block_q4_3(const void * vx, float * y) {
     }
 }
 
-extern "C" {
-    __host__ void dequantize_row_q4_0_cuda(const void * vx, float * y, int k, cudaStream_t stream) {
-        const int nb = k / QK4_0;
-        dequantize_block_q4_0<<<nb, 1, 0, stream>>>(vx, y);
-    }
+void dequantize_row_q4_0_cuda(const void * vx, float * y, int k, cudaStream_t stream) {
+    const int nb = k / QK4_0;
+    dequantize_block_q4_0<<<nb, 1, 0, stream>>>(vx, y);
+}
 
-    __host__ void dequantize_row_q4_1_cuda(const void * vx, float * y, int k, cudaStream_t stream) {
-        const int nb = k / QK4_1;
-        dequantize_block_q4_1<<<nb, 1, 0, stream>>>(vx, y);
+void dequantize_row_q4_1_cuda(const void * vx, float * y, int k, cudaStream_t stream) {
+    const int nb = k / QK4_1;
+    dequantize_block_q4_1<<<nb, 1, 0, stream>>>(vx, y);
+}
+
+void dequantize_row_q4_2_cuda(const void * vx, float * y, int k, cudaStream_t stream) {
+    const int nb = k / QK4_2;
+    dequantize_block_q4_2<<<nb, 1, 0, stream>>>(vx, y);
+}
+
+void dequantize_row_q4_3_cuda(const void * vx, float * y, int k, cudaStream_t stream) {
+    const int nb = k / QK4_3;
+    dequantize_block_q4_3<<<nb, 1, 0, stream>>>(vx, y);
+}
+
+// lock-free, thread safe buffer pool for cuda
+#define MAX_CUDA_BUFFERS 16
+struct cuda_buffer {
+    std::atomic_uintptr_t ptr { 0 };
+    size_t size { 0 };
+};
+
+static cuda_buffer cuda_buffer_pool[MAX_CUDA_BUFFERS];
+
+void * ggml_cuda_pool_malloc(size_t size, size_t * actual_size) {
+    for (int i = 0; i < MAX_CUDA_BUFFERS; ++i) {
+        struct cuda_buffer * b = &cuda_buffer_pool[i];
+        if (b->size >= size) {
+            uintptr_t ptr = atomic_load(&b->ptr);
+            if (ptr) {
+                if (std::atomic_compare_exchange_strong(&b->ptr, &ptr, 0)) {
```

Comment:
Doesn't this has an ABA problem? The scenario I'm thinking of goes like this:
  * we get preempted just before the CAS and lose the race to `ptr` to another thread
  * this other thread eventually frees `ptr` with `cudaFree()`
  * some other thread calls `cudaMalloc()` with a smaller size and get the same pointer as `ptr` (i.e., it is equal to it as an integer), then frees it into the same pool slot we are trying to use
  * we wake up, do the CAS (which succeeds because the new pointer is equal to the old one as integer) and start using the pointer with the wrong size

This is highly unlikely to happen in practice, but I think is technically possible, unless CUDART never returns the same pointer twice from `cudaMalloc()`.